src/v/compat/run.cc

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "compat/run.h"
+
+#include "compat/check.h"
+#include "json/document.h"
+#include "json/prettywriter.h"
+#include "json/writer.h"
+#include "seastarx.h"
+#include "utils/base64.h"
+#include "utils/file_io.h"
+
+#include <seastar/core/thread.hh>
+
+namespace compat {
+
+template<typename T>
+struct corpus_helper {
+    using checker = compat_check<T>;
+
+    /*
+     * Builds a test case for an instance of T.
+     *
+     * {
+     *   "name": "raft::some_request",
+     *   "fields": { test case field values },
+     *   "binaries": [
+     *     {
+     *       "name": "serde",
+     *       "data": "base64 encoding",
+     *     },
+     *   ]
+     * }
+     */
+    static void write_test_case(T t, json::Writer<json::StringBuffer>& w) {
+        auto&& [ta, tb] = compat_copy(std::move(t));
+
+        w.StartObject();
+        w.Key("name");
+        w.String(checker::name.data());
+
+        w.Key("fields");
+        w.StartObject();
+        checker::to_json(std::move(ta), w);
+        w.EndObject();
+
+        w.Key("binaries");
+        w.StartArray();
+        for (auto& b : checker::to_binary(std::move(tb))) {
+            w.StartObject();
+            w.Key("name");
+            w.String(b.name);
+            w.Key("data");
+            w.String(iobuf_to_base64(b.data));
+            w.EndObject();
+        }
+        w.EndArray();
+        w.EndObject();
+    }
+
+    /*
+     * Writes all test cases to the provided directory.
+     */
+    static ss::future<> write(const std::filesystem::path& dir) {
+        size_t instance = 0;
+
+        for (auto& test : checker::create_test_cases()) {
+            // json encoded test case
+            auto buf = json::StringBuffer{};
+            auto writer = json::Writer<json::StringBuffer>{buf};
+            write_test_case(std::move(test), writer);
+
+            // save test case to file
+            iobuf data;
+            data.append(buf.GetString(), buf.GetSize());
+            auto fn = fmt::format("{}_{}.json", checker::name, instance++);
+            write_fully(dir / fn, std::move(data)).get();
```

Comment:
Since AFAIK none of this really depends on seastar could we avoid using methods that return a future and just write to file using the stdlib / boost? It would mean the corpus project wouldn't have to link against seastar, IMO minimizing the space of what we are testing and what could go wrong.

Reply:
I know there is a `serde::read/write_async` but does that follow a different codepath then `serde::read/write`?

Reply:
i don't think we can. i didn't futurize the interfaces yet, but some stuff in raft uses `adl/serde::async` which is going to be viral

Reply:
ah alright in that case no biggie, resolving this